util/containerutil/frontend_test.go

```diff
@@ -592,23 +594,96 @@ func isBinaryInstalled(ctx context.Context, binary string) bool {
+// Then we return
+// Caller is always expected to call cleanup
 func spawnTestContainers(ctx context.Context, feBinary string, names ...string) (func(), error) {
+	removeContainers(ctx, feBinary, names...) // best effort
+	err := startTestContainers(ctx, feBinary, names...)
+	cleanup := func() {
+		removeContainers(ctx, feBinary, names...) // best-effort
+	}
+	if err != nil {
+		return cleanup, err
+	}
+	err = waitForContainers(ctx, feBinary, names...)
+	return cleanup, err
+}
+
+func startTestContainers(ctx context.Context, feBinary string, names ...string) error {
 	var err error
+	m := sync.Mutex{}
+	wg := sync.WaitGroup{}
 	for _, name := range names {
-		cmd := exec.CommandContext(ctx, feBinary, "run", "-d", "--name", name, "docker.io/library/nginx:1.21", `-text="test"`)
-		output, createErr := cmd.CombinedOutput()
-		if createErr != nil {
-			// the frontend exists but is non-functional. This is... not likely to work at all.
-			err = multierror.Append(err, errors.Wrap(createErr, string(output)))
-		}
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			cmd := exec.CommandContext(ctx, feBinary, "run", "-d", "--rm", "--name", name, "docker.io/library/nginx:1.21", "sh", "-c", `echo output stream&&>&2 echo error stream&&sleep 100`)
+			output, createErr := cmd.CombinedOutput()
+			m.Lock()
+			defer m.Unlock()
+			if createErr != nil {
+				// the frontend exists but is non-functional. This is... not likely to work at all.
+				err = multierror.Append(err, errors.Wrap(createErr, string(output)))
+			}
+		}(name)
+	}
+	wg.Wait()
+	return err
+}
+
+func removeContainers(ctx context.Context, feBinary string, names ...string) error {
+	var err error
+	m := sync.Mutex{}
+	wg := sync.WaitGroup{}
+	for _, name := range names {
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			removeCmd := exec.CommandContext(ctx, feBinary, "rm", "-f", name)
+			output, removeErr := removeCmd.CombinedOutput()
+			m.Lock()
+			defer m.Unlock()
+			if removeErr != nil {
+				err = multierror.Append(err, fmt.Errorf("failed to remove container %s", name))
+				fmt.Printf("Warning: failed to remove container with name %s: err: %s, output: %s", name, removeErr, output)
+			}
+		}(name)
 	}
+	wg.Wait()
+	return err
+}
 
-	return func() {
-		for _, name := range names {
-			cmd := exec.CommandContext(ctx, feBinary, "rm", "-f", name)
-			_ = cmd.Run() // Just best effort
-		}
-	}, nil
+func waitForContainers(ctx context.Context, feBinary string, names ...string) error {
+	var err error
+	m := sync.Mutex{}
+	wg := sync.WaitGroup{}
+	for _, name := range names {
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			attempts := 0
+			for attempts < 100 {
+				attempts++
+				// docker inspect -f {{.State.Running}} CONTAINERNAME`"=="true"
+				cmd := exec.CommandContext(ctx, feBinary, "inspect", "-f", "{{.State.Running}}", name)
+				output, _ := cmd.CombinedOutput()
+				if strings.Contains(string(output), "true") {
+					break
```

Comment:
might be best to just return here instead,

then if you ever get to the end of the `for` loop, you'll know there's an error.

this would make things more robust since the for loop references `100` attempts, but the error checking below references `1000`.

otherwise, it's probably better to use a maxAttempts const.

Reply:
Good catch! I missed that in the refactor.
I'm using a return now and a maxAttempts const.